### PR TITLE
clean up solution type tracking

### PR DESCRIPTION
**Pull request by smichr**
#### Brief description of what is fixed or changed

Consolidated code and improved solution type handling in `solve`

#### Release Notes

* solvers
  * solutions for lists of nonlinear equations should no longer have duplicate solutions
  * multiple solutions for nonlinear equations for which symbols have been specified should more consistently return as tuples
  * dictionary solutions are sorted consistently by *key* not by item
  * when using set=True, all symbols specified will be returned in the order given (or canonical order if unordered); if no symbols are given then only those symbols isolated will be included in the list of variables

### CODE REVIEW

File: sympy/solvers/solvers.py [hunk 1]
```diff
@@ -881,11 +877,20 @@ def _sympified_list(w):
...
+        if len(symbols) == 1 and iterable(symbols[0]):
+            symbols = symbols[0]
+        ordered_symbols = symbols and is_sequence(symbols,
+                        include=GeneratorType)
+        _symbols = list(uniq(symbols))
+        if len(_symbols) != len(symbols):
+            ordered_symbols = False
+            symbols = list(ordered(symbols))
+        else:
+            symbols = _symbols
```
Comment (smichr) [inline sympy/solvers/solvers.py:892]:
```suggestion
        ordered_symbols = False
        if symbols:
            if is_sequence(symbols, include=GeneratorType):
                # the symbols are well ordered; keep the
                # order but make sure they are unique
                _symbols = list(uniq(symbols))
                if len(_symbols) == len(symbols):
                    # the symbols are unique so we can return
                    # solutions relative to their given order
                    ordered_symbols = True
                symbols = _symbols
            else:
                # put them in order for canonical processing
                symbols = list(ordered(symbols))
                if flags.get('set', False):
                    # since symbols were provided, all will be
                    # included in the set output; even though they
                    # were not well ordered, they are now
                    ordered_symbols = True
```

**Comment by oscarbenjamin** [comment]
I find it very hard to follow the logic of what is happening here. This seems to be an improvement but it's difficult to follow all cases in the control flow.

It would be a lot better it `_solve` and `_solve_system` always returned a consistent type e.g. list of dict. Then there wouldn't be so many cases to be handled in the code after calling them. Converting to the output form that `solve` is currently expected to return should really be the last thing that happens in `solve` and internally at least the types can be consistent.

I'm not suggesting that this PR should necessarily fix that but it would be a lot easier to understand changes like this if that was fixed. Maybe it's not even possible to express the logic of how `solve` determines its output types that way though (i.e. the output types are governed by what `_solve` and `_solve_system` return).

**Comment by smichr** [comment]
> I'm not suggesting that this PR should necessarily fix that but

I'm trying to take this in smaller steps. 
* This is `s1`. The subtle change here is that a case that should have come back as tuple instead of a dict in a list has been fixed. Also there is a commitment to include all symbols requested in the `set` output form whether those symbols appear in the equations or not -- as given (if given in an ordered way) or else canonically ordered (if given as a set).
* In `s2` the nonlinear coefficient system is no longer handled
* In `s3` the `solve_output.rst` will be added to Explanations.
* Then I am waiting for `has_xfree` before starting `s4` to modify `solve_undetermined_coeffs`. 
* `s5` will fold all return types from the `solve` step into a list of dicts, post process, and unfold at point of return. 
* `s6` will be a major overhaul of the manual fallback and solve many issues that we currently have.

**Comment by oscarbenjamin** [comment]
Okay, I think this looks good. Honestly I can't follow all of the control flow but I've checked the tests and they all look like improvements in consistency. The control flow looks like it's becoming more complicated but +1 to merge this if that is going to be simplified later because this seems to be an improvement in correctness.

The release notes should clarify that they are talking about `solve`.

> * `s5` will fold all return types from the `solve` step into a list of dicts, post process, and unfold at point of return.

I think I would approach this the other way around. My step 1 would be to find all of the places where `_solve` and `_solve_system` generate something that is not in list-of-dict form and convert the result immediately into a list of dicts. That way we handle the different formats at source and `solve` can always *receive* a list of dicts.
